### DISCUSSION TITLE
How to hot update TLS certificate and do not impact existing TLS connections?

### DISCUSSION THREAD

**Discussion by duxiaocheng** [openssl/openssl#23185] [Q&A] [answered]
Hello Everyone,

I want to hot update TLS certificate, but I do not want to impact existing TLS connections. That means those existing TLS connection still use old TLS certificate and keep them there. New TLS connections should use that new TLS certificate for TLS handshake. Below is my initial idea. Not sure if it is safe or not. What is the best practice? Welcome your comments! Thanks!

Existing SSL CTX and SSL/TLS connection.
```
// Create new SSL_CTX
SSL_CTX *g_ctx = SSL_CTX_new();
// Set the initial TLS certificate file
SSL_CTX_use_certificate_chain_file(g_ctx , cert_file);
...
SSL *p_ssl_1 = SSL_new(g_ctx);
// create TLS connection with p_ssl_1 
...
```

Hot update TLS certificate file
```
// Remove that old SSL_CTX and use new TLS certificate to initialize it
SSL_CTX_free(g_ctx );
g_ctx = SSL_CTX_new();
SSL_CTX_use_certificate_chain_file(g_ctx , new_cert_file);
...
SSL *p_ssl_2 = SSL_new(g_ctx);
// create new TLS connection with p_ssl_2
...
```
Finally, there are two TLS connections: one is using old TLS certificate p_ssl_1 and another one is using new TLS certificate p_ssl_2.

But, I saw below note from openssl manual. I am wondering if I am wrong to free old SSL_CTX and create a new one?
https://www.openssl.org/docs/manmaster/man3/SSL_CTX_new.html
An SSL_CTX object should not be changed after it is used to create any SSL objects or from multiple threads concurrently, since the implementation does not provide serialization of access for these cases.

It is appreciated that you can give some suggestions. Thanks!

Best Regards,
Cheng

**Comment by t8m** [comment, marked as answer]
SSL_CTX objects are refcounted and the reference used in SSL_new() is counted so it is possible to call SSL_CTX_free() on a SSL_CTX object before any SSL objects created from the context are freed. The SSL_CTX will be properly destroyed when the last SSL object referencing it is freed.

**Reply by duxiaocheng** [reply]
Thank you for your reply!
So, it is ok to call SSL_CTX_free() before SSL_free(). Right?

Another implementation is to call SSL_CTX_use_certificate_chain_file/SSL_CTX_use_PrivateKey_file/SSL_CTX_load_verify_locations to update that old SSL_CTX with new certificate/ca files. I think this is also ok. Which one is the best practice?

https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_use_certificate_chain_file.html
The SSL_CTX_* class of functions loads the certificates and keys into the SSL_CTX object ctx. The information is passed to SSL objects ssl created from ctx with [SSL_new(3)](https://www.openssl.org/docs/man1.1.1/man3/SSL_new.html) by copying, so that changes applied to ctx do not propagate to already existing SSL objects.

**Reply by mattcaswell** [reply]
> So, it is ok to call SSL_CTX_free() before SSL_free(). Right?

Yes, this is ok.

> Another implementation is to call SSL_CTX_use_certificate_chain_file/SSL_CTX_use_PrivateKey_file/SSL_CTX_load_verify_locations to update that old SSL_CTX with new certificate/ca files. I think this is also ok. Which one is the best practice?

This modifies the SSL_CTX which is already in use. Don't do this.

> The SSL_CTX_* class of functions loads the certificates and keys into the SSL_CTX object ctx. The information is passed to SSL objects ssl created from ctx with [SSL_new(3)](https://www.openssl.org/docs/man1.1.1/man3/SSL_new.html) by copying, so that changes applied to ctx do not propagate to already existing SSL objects.

You should not rely on this. This is not guaranteed to remain stable behaviour.

**Reply by duxiaocheng** [reply]
Thank you very much for your detailed reply.
